connor/connor.go

```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		return nil, err
+	}
+
+	balanceReply, err := connor.TokenClient.Balance(ctx, &sonm.Empty{})
+	if err != nil {
+		connor.logger.Error("cannot load balanceReply", zap.Error(err))
+		return nil, err
+	}
+
+	connor.logger = ctxlog.GetLogger(ctx)
+	connor.logger.Info("config",
+		zap.String("eth node address", cfg.Market.Endpoint),
+		zap.String("key", crypto.PubkeyToAddress(key.PublicKey).String()))
+	connor.logger.Info("balance",
+		zap.String("live", balanceReply.GetLiveBalance().Unwrap().String()),
+		zap.String("Side", balanceReply.GetSideBalance().ToPriceString()))
+	connor.logger.Info("configuring connor", zap.Any("config", cfg))
+	return connor, nil
+}
+
+func (c *Connor) Serve(ctx context.Context) error {
+	c.logger.Info("Connor started work ...")
+	defer c.logger.Info("Connor has been stopped")
+
+	c.ClearStart(ctx)
+
+	dataUpdate := time.NewTicker(10 * time.Second)
+	defer dataUpdate.Stop()
+	tradeUpdate := time.NewTicker(15 * time.Second)
+	defer tradeUpdate.Stop()
+	poolInit := time.NewTicker(990 * time.Second)
+	defer poolInit.Stop()
+
+	snm := watchers.NewSNMPriceWatcher(coinMarketCapSonmTicker)
+	token := watchers.NewTokenPriceWatcher(coinMarketCapTicker, cryptoCompareCoinData)
+	reportedPool := watchers.NewPoolWatcher(poolReportedHashRate, []string{c.cfg.PoolAddress.EthPoolAddr})
+	avgPool := watchers.NewPoolWatcher(poolAverageHashRate, []string{c.cfg.PoolAddress.EthPoolAddr + "/1"})
+
+	if err := snm.Update(ctx); err != nil {
+		return fmt.Errorf("cannot update snm data: %v", err)
+	}
+	if err := token.Update(ctx); err != nil {
+		return fmt.Errorf("cannot update token data: %v", err)
+	}
+	if err := reportedPool.Update(ctx); err != nil {
+		return fmt.Errorf("cannot update reportedPool data: %v", err)
+	}
+	if err := avgPool.Update(ctx); err != nil {
+		return fmt.Errorf("cannot update avgPool data: %v", err)
+	}
+
+	profitModule := NewProfitableModules(c)
+	poolModule := NewPoolModules(c)
+	traderModule := NewTraderModules(c, poolModule, profitModule)
+
+	balanceReply, err := c.TokenClient.Balance(ctx, &sonm.Empty{})
+	if err != nil {
+		return err
+	}
+	md := errgroup.Group{}
+	md.Go(func() error {
+		return traderModule.ChargeOrdersOnce(ctx, c.cfg.UsingToken.Token, token, snm, balanceReply)
+	})
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context done %v", ctx.Err())
+		case <-dataUpdate.C:
+			if err := snm.Update(ctx); err != nil {
+				return fmt.Errorf(" cannot update SNM data: %v\n", err)
+			}
+			if err := token.Update(ctx); err != nil {
+				return fmt.Errorf("cannot update TOKEN data: %v\n", err)
+			}
+			if err := reportedPool.Update(ctx); err != nil {
+				return fmt.Errorf("cannot update reported pool data: %v\n", err)
+			}
+			if err := avgPool.Update(ctx); err != nil {
+				return fmt.Errorf("cannot update avg pool data: %v\n", err)
+			}
+			go profitModule.CollectTokensMiningProfit(token)
```

Comment:
I have doubs about this. Since you spawn some work within a loop which itself is controlled by a ctx could we control things with a wg or something and handle termination in a more clean way?